src/main/java/com/mraof/minestuck/world/gen/feature/structure/tiered/tier1/TierOneDungeonSecondaryPiece.java

```diff
@@ -0,0 +1,638 @@
+	private BlockState secondaryBlock;
+	private BlockState secondaryDecorativeBlock;
+	private BlockState lightBlock;
+	
+	private EnumAspect worldAspect;
+	private EnumClass worldClass;
+	private TerrainLandType worldTerrain;
+	private Template breathSymbolTemplate;
+	private Template breathPuzzleDoorTemplate;
+	private Template breathFirstSideRoomTemplate;
+	private Template breathSecondSideRoomBarrierTemplate;
+	private Template breathAreaEffectChamberTemplate;
+	private Template bloodSymbolTemplate;
+	private Template bloodFirstSideRoomTemplate;
+	private Template bloodSecondSideRoomTemplate;
+	private Template bloodWallFountainTemplate;
+	
+	public TierOneDungeonSecondaryPiece(TemplateManager templates, Direction direction, int x, int y, int z) //this constructor is used when the structure is first initialized
+	{
+		super(MSStructurePieces.TIER_ONE_DUNGEON_SECONDARY, 0);
+		
+		setCoordBaseMode(direction);
+		setBounds(x, y, z, 82, 50, 82);
+		
+		initTemplates(templates);
+	}
+	
+	public TierOneDungeonSecondaryPiece(TemplateManager templates, CompoundNBT nbt) //this constructor is used for reading from data
+	{
+		super(MSStructurePieces.TIER_ONE_DUNGEON_SECONDARY, nbt);
+		spawner1 = nbt.getBoolean("sp1");
+		spawner2 = nbt.getBoolean("sp2");
+		randomRoomType = nbt.getInt("randomRoomType");
+		roomVariable1 = nbt.getInt("roomVariable1");
+		
+		initTemplates(templates);
+	}
+	
+	@Override
+	protected void readAdditional(CompoundNBT tagCompound) //actually writeAdditional
+	{
+		tagCompound.putBoolean("sp1", spawner1); //spawner type room only
+		tagCompound.putBoolean("sp2", spawner2); //spawner type room only
+		tagCompound.putInt("randomRoomType", randomRoomType);
+	}
+	
+	private void initTemplates(TemplateManager templates)
+	{
+		breathSymbolTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "aspect/breath_symbol_no_background"));
+		breathPuzzleDoorTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/breath_puzzle_door"));
+		breathFirstSideRoomTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/breath_first_side_room"));
+		breathSecondSideRoomBarrierTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/breath_second_side_room_barrier"));
+		breathAreaEffectChamberTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/breath_area_effect_chamber"));
+		bloodSymbolTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "aspect/blood_symbol_no_background"));
+		bloodFirstSideRoomTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/blood_first_side_room"));
+		bloodSecondSideRoomTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/blood_second_side_room"));
+		bloodWallFountainTemplate = templates.getOrCreate(new ResourceLocation(Minestuck.MOD_ID, "dungeons/blood_wall_fountain")); //getOrCreate was getTemplateDefaulted
+	}
```

Comment:
Considering how many lines of code will be needed for the 12+ dungeon templates I think this init system should be changed